pkg/envtest/apiserver.go

```diff
@@ -0,0 +1,443 @@
+		if err != nil {
+			return fmt.Errorf("failed checking for gardener-apiserver binary under %q: %w", a.Path, err)
+		}
+		log.V(1).Info("using pre-built gardener-apiserver test binary", "path", a.Path)
+	}
+
+	if a.SecurePort == 0 {
+		a.SecurePort, _, err = SuggestPort("")
+		if err != nil {
+			return err
+		}
+	}
+
+	// resolve localhost IP (pin to IPv4)
+	addr, err := net.ResolveTCPAddr("tcp", net.JoinHostPort("localhost", "0"))
+	if err != nil {
+		return err
+	}
+	a.listenURL = &url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(addr.IP.String(), strconv.Itoa(a.SecurePort)),
+	}
+
+	if a.HealthCheckEndpoint == "" {
+		a.HealthCheckEndpoint = "/healthz"
+	}
+
+	kubeconfigFile, err := a.prepareKubeconfigFile()
+	if err != nil {
+		return err
+	}
+
+	a.Args = append([]string{
+		"--bind-address=" + addr.IP.String(),
+		"--etcd-servers=" + a.EtcdURL.String(),
+		"--tls-cert-file=" + filepath.Join(a.CertDir, "tls.crt"),
+		"--tls-private-key-file=" + filepath.Join(a.CertDir, "tls.key"),
+		"--secure-port=" + fmt.Sprintf("%d", a.SecurePort),
+		"--cluster-identity=envtest",
+		"--authorization-always-allow-paths=" + a.HealthCheckEndpoint,
+		"--authentication-kubeconfig=" + kubeconfigFile,
+		"--authorization-kubeconfig=" + kubeconfigFile,
+		"--kubeconfig=" + kubeconfigFile,
+	}, a.Args...)
+
+	return nil
+}
+
+// prepareKubeconfigFile marshals the test environments rest config to a kubeconfig file in the CertDir.
+func (a *GardenerAPIServer) prepareKubeconfigFile() (string, error) {
+	kubeconfigBytes, err := util.MarshalKubeconfigWithClientCertificate(a.restConfig, nil, nil)
+	if err != nil {
+		return "", err
+	}
+	kubeconfigFile := filepath.Join(a.CertDir, "kubeconfig.yaml")
+
+	return kubeconfigFile, ioutil.WriteFile(kubeconfigFile, kubeconfigBytes, 0600)
+}
+
+// waitUntilHealthy waits for the HealthCheckEndpoint to return 200.
+func (a *GardenerAPIServer) waitUntilHealthy(ctx context.Context) error {
+	// setup secure http client
+	certPool := x509.NewCertPool()
+	certPool.AppendCertsFromPEM(a.caCert.CertificatePEM)
+	httpClient := &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{RootCAs: certPool}}}
+
+	healthCheckURL := a.listenURL
+	healthCheckURL.Path = a.HealthCheckEndpoint
+
+	err := retry.Until(ctx, waitPollInterval, func(context.Context) (bool, error) {
+		res, err := httpClient.Get(healthCheckURL.String())
+		if err == nil {
+			_ = res.Body.Close()
+			if res.StatusCode == http.StatusOK {
+				log.V(1).Info("gardener-apiserver got healthy")
+				return retry.Ok()
+			}
+		}
+		return retry.MinorError(err)
+	})
+	if err != nil {
+		if stopErr := a.Stop(); stopErr != nil {
+			log.Error(stopErr, "failed stopping gardener-apiserver")
+		}
+	}
+	return err
+}
+
+var allGardenerAPIGroupVersions = []schema.GroupVersion{
+	gardencorev1beta1.SchemeGroupVersion,
+	gardencorev1alpha1.SchemeGroupVersion,
+	settingsv1alpha1.SchemeGroupVersion,
+	seedmanagementv1alpha1.SchemeGroupVersion,
+}
```

Comment:
Minor: Does it make sense to maintain this here: https://github.com/gardener/gardener/tree/master/pkg/api?

Reply:
I was also thinking about putting this list in a central place.
Though, I found `pkg/api` and `pkg/apis` to be odd, as they also host types/helpers for the extensions API group, which is not hosted by g-api.
I moved it to `pkg/apiserver` now, hope this is a good fit.